PES2UG23CS072

In [8]:
from transformers import pipeline

In [9]:
print("--- Experiment 1: Text Generation ---\n")

# 1. BERT
try:
    gen_bert = pipeline('text-generation', model='bert-base-uncased')
    print("BERT Result:", gen_bert("The future of Artificial Intelligence is"))
except Exception as e:
    print(f"BERT Failed: {e}")

# 2. RoBERTa
try:
    gen_roberta = pipeline('text-generation', model='roberta-base')
    print("RoBERTa Result:", gen_roberta("The future of Artificial Intelligence is"))
except Exception as e:
    print(f"RoBERTa Failed: {e}")

# 3. BART
try:
    gen_bart = pipeline('text-generation', model='facebook/bart-base')
    print("BART Result:", gen_bart("The future of Artificial Intelligence is", max_length=20))
except Exception as e:
    print(f"BART Failed: {e}")

If you want to use `BertLMHeadModel` as a standalone, add `is_decoder=True.`


--- Experiment 1: Text Generation ---



Device set to use cuda:0
If you want to use `RobertaLMHeadModel` as a standalone, add `is_decoder=True.`


BERT Result: [{'generated_text': 'The future of Artificial Intelligence is................................................................................................................................................................................................................................................................'}]


Device set to use cuda:0


RoBERTa Result: [{'generated_text': 'The future of Artificial Intelligence is'}]


Some weights of BartForCausalLM were not initialized from the model checkpoint at facebook/bart-base and are newly initialized: ['lm_head.weight', 'model.decoder.embed_tokens.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
Device set to use cuda:0
Truncation was not explicitly activated but `max_length` is provided a specific value, please use `truncation=True` to explicitly truncate examples to max length. Defaulting to 'longest_first' truncation strategy. If you encode pairs of sequences (GLUE-style) with the tokenizer you can select this strategy more precisely by providing a specific strategy to `truncation`.
Both `max_new_tokens` (=256) and `max_length`(=20) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


BART Result: [{'generated_text': 'The future of Artificial Intelligence is� Middles Middles finding PS SuicideBehind chemicals�Friend Suicide finding finding finding FerrariIntroduIntrodu Verb Orbit Middles Middlesrationsloaderacusrationsrations Baldwinpriv Middlesrations rulrations finding findingwomanwomanloaderPaulexternalActionCode felonrationsexternalActionCodeexternalActionCodeFriendFYFYexternalActionCodeexternalActionCodeloader frayFriendFriendFriendandelanciescks abducted abducted abducted DidnanciesrationsFriendFriend abductedBadrationsrationsexternalActionCodetipsktopanciestips fray abducted hemispherewomanrationsancies abducted abducted{"rations paving abductedexternalActionCode defects SuiciderationsrationsBad abductedrations defectsancies abducted nobleloaderloaderrationsrations fighterwoman abducted abductedanciesrations abductedFriendcksancies{" abducted{" abducted abducted vileFriend abducted Baldwin abducted abducted help abducted noble abducted Channelanciesrationsrat

In [ ]:
print("\n--- Experiment 2: Fill-Mask ---\n")

# 1. BERT
unmasker_bert = pipeline('fill-mask', model='bert-base-uncased')
print("BERT:", unmasker_bert("The goal of Generative AI is to [MASK] new content."))

# 2. RoBERTa
unmasker_roberta = pipeline('fill-mask', model='roberta-base')
print("RoBERTa:", unmasker_roberta("The goal of Generative AI is to <mask> new content."))

# 3. BART
unmasker_bart = pipeline('fill-mask', model='facebook/bart-base')
print("BART:", unmasker_bart("The goal of Generative AI is to <mask> new content."))

Some weights of the model checkpoint at bert-base-uncased were not used when initializing BertForMaskedLM: ['bert.pooler.dense.bias', 'bert.pooler.dense.weight', 'cls.seq_relationship.bias', 'cls.seq_relationship.weight']
- This IS expected if you are initializing BertForMaskedLM from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertForMaskedLM from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).



--- Experiment 2: Fill-Mask ---



Device set to use cuda:0


BERT: [{'score': 0.5396888852119446, 'token': 3443, 'token_str': 'create', 'sequence': 'the goal of generative ai is to create new content.'}, {'score': 0.15575668215751648, 'token': 9699, 'token_str': 'generate', 'sequence': 'the goal of generative ai is to generate new content.'}, {'score': 0.054054468870162964, 'token': 3965, 'token_str': 'produce', 'sequence': 'the goal of generative ai is to produce new content.'}, {'score': 0.04451529309153557, 'token': 4503, 'token_str': 'develop', 'sequence': 'the goal of generative ai is to develop new content.'}, {'score': 0.01757732406258583, 'token': 5587, 'token_str': 'add', 'sequence': 'the goal of generative ai is to add new content.'}]


Device set to use cuda:0


RoBERTa: [{'score': 0.3711293935775757, 'token': 5368, 'token_str': ' generate', 'sequence': 'The goal of Generative AI is to generate new content.'}, {'score': 0.36771273612976074, 'token': 1045, 'token_str': ' create', 'sequence': 'The goal of Generative AI is to create new content.'}, {'score': 0.08351442217826843, 'token': 8286, 'token_str': ' discover', 'sequence': 'The goal of Generative AI is to discover new content.'}, {'score': 0.021335095167160034, 'token': 465, 'token_str': ' find', 'sequence': 'The goal of Generative AI is to find new content.'}, {'score': 0.016521504148840904, 'token': 694, 'token_str': ' provide', 'sequence': 'The goal of Generative AI is to provide new content.'}]


Device set to use cuda:0


BART: [{'score': 0.0746147632598877, 'token': 1045, 'token_str': ' create', 'sequence': 'The goal of Generative AI is to create new content.'}, {'score': 0.06571780890226364, 'token': 244, 'token_str': ' help', 'sequence': 'The goal of Generative AI is to help new content.'}, {'score': 0.060879286378622055, 'token': 694, 'token_str': ' provide', 'sequence': 'The goal of Generative AI is to provide new content.'}, {'score': 0.03593532741069794, 'token': 3155, 'token_str': ' enable', 'sequence': 'The goal of Generative AI is to enable new content.'}, {'score': 0.03319435939192772, 'token': 1477, 'token_str': ' improve', 'sequence': 'The goal of Generative AI is to improve new content.'}]


In [11]:
print("\n--- Experiment 3: QA ---\n")

context = "Generative AI poses significant risks such as hallucinations, bias, and deepfakes."
question = "What are the risks?"

# 1. BERT
qa_bert = pipeline('question-answering', model='bert-base-uncased')
print("BERT:", qa_bert(question=question, context=context))

# 2. RoBERTa
qa_roberta = pipeline('question-answering', model='roberta-base')
print("RoBERTa:", qa_roberta(question=question, context=context))

# 3. BART
qa_bart = pipeline('question-answering', model='facebook/bart-base')
print("BART:", qa_bart(question=question, context=context))

Some weights of BertForQuestionAnswering were not initialized from the model checkpoint at bert-base-uncased and are newly initialized: ['qa_outputs.bias', 'qa_outputs.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.



--- Experiment 3: QA ---



Device set to use cuda:0
Some weights of RobertaForQuestionAnswering were not initialized from the model checkpoint at roberta-base and are newly initialized: ['qa_outputs.bias', 'qa_outputs.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


BERT: {'score': 0.011016791686415672, 'start': 72, 'end': 82, 'answer': 'deepfakes.'}


Device set to use cuda:0
Some weights of BartForQuestionAnswering were not initialized from the model checkpoint at facebook/bart-base and are newly initialized: ['qa_outputs.bias', 'qa_outputs.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


RoBERTa: {'score': 0.013165182434022427, 'start': 72, 'end': 81, 'answer': 'deepfakes'}


Device set to use cuda:0


BART: {'score': 0.046605834271758795, 'start': 68, 'end': 81, 'answer': 'and deepfakes'}


| Task | Model | Classification (Success/Failure) | Observation (What actually happened?) | Why did this happen? (Architectural Reason) |
| :--- | :--- | :--- | :--- | :--- |
| **Generation** | BERT | **Failure** | Output a long string of repeated dots (`........`). | BERT is an **Encoder**. It uses bi-directional attention (seeing all words at once) and is not trained to predict the next word in a sequence (auto-regression). |
| | RoBERTa | **Failure** | Output the prompt exactly and stopped (`The future of...`). | Like BERT, RoBERTa is an **Encoder** only. It lacks a decoder component to generate new tokens and was not trained for next-token prediction. |
| | BART | **Failure** (Quality) | Generated incoherent gibberish (`Middles Middles finding...`). | While BART is an **Encoder-Decoder** (capable of generation), the *base* model is trained on **Denoising** (reconstructing text), not **Causal LM** (continuing text). The weights for open-ended generation were not initialized. |
| **Fill-Mask** | BERT | **Success** | Predicted `create` (54%) and `generate` (15%). | BERT is natively trained on **Masked Language Modeling (MLM)**. This is its primary training objective. |
| | RoBERTa | **Success** | Predicted `generate` (37%) and `create` (36%). | RoBERTa is also trained on MLM (specifically with dynamic masking), making it highly effective at this task. |
| | BART | **Success** | Predicted `create` (7%) and `help` (6%). | BART is trained to reconstruct corrupted text (Denoising). Filling in a missing word (`<mask>`) is a form of text reconstruction, which it excels at. |
| **QA** | BERT | **Random / Poor** | Extracted `deepfakes.` but with very low confidence (`0.01`). | The base model has the **Architecture** (Encoder) to extract spans, but lacks the **Fine-tuning** (SQuAD training) to do it accurately. The "Head" weights were random. |
| | RoBERTa | **Random / Poor** | Extracted `deepfakes` with very low confidence (`0.01`). | Same as BERT. It is a powerful Encoder that can "read" the text, but without QA fine-tuning, it doesn't know *which* span is the answer to the question. |
| | BART | **Random / Poor** | Extracted `and deepfakes` with very low confidence (`0.04`). | BART can perform QA, but like the others, the `base` version is not fine-tuned for it. It just guessed a salient noun phrase from the context. |